### ISSUE TITLE
feat(evaluation): synth matching CLI — predict params from audio and apply to Surge XT

### ISSUE THREAD

**Issue by ktinubu** [tinaudio/synth-setter#491]
## Summary

End-user CLI tool for synth matching: take an audio file, predict Surge XT parameters, and apply them. Two tiers: simple Makefile commands for producers, full CLI flags for researchers.

## End-to-End Flow

```
1. Prerequisites check
   └── weights downloaded? stats.npz present?

2. Weight auto-download (first run only)
   └── GitHub Release → ~/.cache/synth-setter/v0.1.0/checkpoint.ckpt + stats.npz

3. Audio preprocessing
   └── WAV → pedalboard AudioFile → resample 44.1kHz → pad/trim 4s
       → make_spectrogram() → normalize with stats.npz mean/std

4. Model inference
   └── load_from_checkpoint → encode mel-spec (AST) → RK4 sample → [-1,1] params

5. Decode params
   └── Scale [-1,1] → [0,1] → param_spec.decode() → named synth + note params

6. Output
   └── JSON params file + param summary to console
   └── (Future: .vstpreset export when we figure out the format)
```

**Key constraint**: Pedalboard can `load_preset` and `set_params` but has **no `save_preset` method**. V1 outputs JSON params + a second command that applies them to Surge via pedalboard. `.vstpreset` export is a follow-up.

### `scripts/match.py` — predict params from audio

```bash
# Producer tier (Makefile):
make match FILE=input.wav

# Researcher tier:
python scripts/match.py input.wav \
  --output params.json \
  --steps 200 \
  --cfg 4.0 \
  --checkpoint path/to/custom.ckpt \
  --stats path/to/stats.npz \
  --device mps
```

Defaults: `--steps 100`, `--cfg 2.0`, `--device` auto-detects (mps > cuda > cpu), `--output` writes to `{input_stem}_match.json`.

### `scripts/apply_match.py` — load params into Surge XT

```bash
# Producer tier:
make apply FILE=input_match.json

# Researcher tier:
python scripts/apply_match.py input_match.json \
  --plugin-path "plugins/Surge XT.vst3" \
  --preset-path presets/surge-base.vstpreset \
  --render output.wav
```

Loads Surge XT via pedalboard, applies the predicted params, optionally renders audio.

## JSON Output Format

```json
{
  "version": "0.1.0",
  "input_file": "input.wav",
  "synth_params": {
    "osc1_shape": 0.42,
    "filter1_cutoff": 0.78
  },
  "note_params": {
    "pitch": 60,
    "note_start_and_end": [0.05, 3.5]
  },
  "raw_vector": [0.15, -0.33]
}
```

## Weight Distribution

- **Ships in GitHub Release**: `checkpoint.ckpt` + `stats.npz`
- **Cache**: `~/.cache/synth-setter/{version}/`
- **Download**: stdlib `urllib.request` from public release URL
- **Bypass**: `--checkpoint path/to/custom.ckpt` skips auto-download

## Design Decisions

| Question | Answer |
|----------|--------|
| Target user | Both producers and researchers, layered CLI |
| Audio input | CLI file path |
| Param output | JSON params + apply helper (two commands) |
| Weight distribution | GitHub Release assets, plain HTTPS auto-download |
| Surge XT dependency | User installs themselves; discovery handled by existing util |
| Install method | `git clone` → `make setup` → `make match FILE=x.wav` |
| Architecture | scripts/ + Makefile (follows existing patterns) |
| Audio loading | Inline in match.py |
| Model loading | `LightningModule.load_from_checkpoint()` |

## Reusable Components

| Need | Existing code | File |
|------|--------------|------|
| Mel-spectrogram | `make_spectrogram()` | `src/data/audio_datamodule.py` |
| Param decode | `ParamSpec.decode()` | `src/data/vst/param_spec.py` |
| Param specs | `param_specs` dict | `src/data/vst/__init__.py` |
| Load plugin | `load_plugin()` | `src/data/vst/core.py` |
| Set params | `set_params()` | `src/data/vst/core.py` |
| Render audio | `render_params()` | `src/data/vst/core.py` |
| Write WAV | `write_wav()` | `src/data/vst/core.py` |
| Model class | `SurgeFlowMatchingModule` | `src/models/surge_flow_matching_module.py` |

## New Files

1. `scripts/match.py` — CLI + weight download + single-file inference
2. `scripts/apply_match.py` — CLI + JSON loading + Surge param application
3. Makefile targets: `match`, `apply`

## Verification

1. Unit tests: weight download (mock HTTP), JSON output format
2. Integration test (`@pytest.mark.slow`): full pipeline with checkpoint + audio
3. Manual: `make match FILE=test.wav` → inspect JSON �� `make apply FILE=test_match.json`

**Comment by ktinubu** [comment]
# Synth Matching CLI — Design Spec

**Epic**: [#491](https://github.com/tinaudio/synth-setter/issues/491)
**Date**: 2026-04-02
**Status**: Draft

## Problem

The project has working inference machinery (model, audio preprocessing, parameter decoding, VST interaction) but no end-user tool that chains them together. A user with a WAV file and a trained model has no way to get Surge XT parameters without manually orchestrating multiple scripts and data formats.

## Solution

Two CLI scripts with Makefile wrappers that provide a layered experience:

- **Producers**: `make match FILE=input.wav` / `make apply FILE=input_match.json`
- **Researchers**: `python scripts/match.py input.wav --steps 200 --cfg 4.0 --device mps`

## End-to-End Flow

```
1. Prerequisites check
   └── weights downloaded? stats.npz present?

2. Weight auto-download (first run only)
   └── GitHub Release → ~/.cache/synth-setter/v0.1.0/checkpoint.ckpt + stats.npz

3. Audio preprocessing
   └── WAV → pedalboard AudioFile → resample 44.1kHz → pad/trim 4s
       → make_spectrogram() → normalize with stats.npz mean/std

4. Model inference
   └── load_from_checkpoint → encode mel-spec (AST) → RK4 sample → [-1,1] params

5. Decode params
   └── Scale [-1,1] → [0,1] → param_spec.decode() → named synth + note params

6. Output
   └── JSON params file + param summary to console
```

## Commands

### `scripts/match.py` — predict params from audio

```bash
# Producer tier:
make match FILE=input.wav

# Researcher tier:
python scripts/match.py input.wav \
  --output params.json \
  --steps 200 \
  --cfg 4.0 \
  --checkpoint path/to/custom.ckpt \
  --stats path/to/stats.npz \
  --device mps
```

| Flag | Default | Description |
|------|---------|-------------|
| `FILE` (positional) | required | Input WAV file path |
| `--output` | `{stem}_match.json` | Output JSON path |
| `--steps` | `100` | ODE solver steps (more = better quality, slower) |
| `--cfg` | `2.0` | Classifier-free guidance strength |
| `--checkpoint` | auto-download | Path to custom .ckpt (bypasses auto-download) |
| `--stats` | auto-download | Path to custom stats.npz |
| `--device` | auto (mps > cuda > cpu) | Compute device |
| `--param-spec` | `surge_xt` | Parameter specification to use |

### `scripts/apply_match.py` — load params into Surge XT

```bash
# Producer tier:
make apply FILE=input_match.json

# Researcher tier:
python scripts/apply_match.py input_match.json \
  --plugin-path "plugins/Surge XT.vst3" \
  --preset-path presets/surge-base.vstpreset \
  --render output.wav
```

| Flag | Default | Description |
|------|---------|-------------|
| `FILE` (positional) | required | Match JSON file from `match.py` |
| `--plugin-path` | required | Path to Surge XT .vst3 |
| `--preset-path` | `presets/surge-base.vstpreset` | Base preset to load before setting params |
| `--render` | none | If set, render audio to this WAV path |
| `--sample-rate` | `44100.0` | Sample rate for rendering |
| `--duration` | `4.0` | Signal duration for rendering |

## JSON Output Format

```json
{
  "version": "0.1.0",
  "input_file": "input.wav",
  "synth_params": {
    "osc1_shape": 0.42,
    "filter1_cutoff": 0.78
  },
  "note_params": {
    "pitch": 60,
    "note_start_and_end": [0.05, 3.5]
  },
  "raw_vector": [0.15, -0.33, 0.88]
}
```

- `synth_params`: decoded parameter names and float values in [0, 1] (raw VST range)
- `note_params`: MIDI note parameters (pitch, timing)
- `raw_vector`: model output in [-1, 1] before decoding (for researchers)

## Weight Distribution

**What ships in the GitHub Release:**
- `checkpoint.ckpt` — model weights (Lightning checkpoint, embeds hparams)
- `stats.npz` — mel-spectrogram normalization statistics (mean, std)

**Cache location:** `~/.cache/synth-setter/{version}/`

**Download mechanism:** stdlib `urllib.request` from the public GitHub Release URL:
```
https://github.com/tinaudio/synth-setter/releases/download/v{version}/{filename}
```

No extra dependencies (no `gh` CLI, no `requests`). Clear error message if offline. The model version is pinned in the script (e.g., `MODEL_VERSION = "v0.1.0"`).

**Bypass:** `--checkpoint path/to/custom.ckpt` and `--stats path/to/stats.npz` skip auto-download.

## Model Loading

Uses Lightning's built-in checkpoint loading:

```python
model = SurgeFlowMatchingModule.load_from_checkpoint(ckpt_path, map_location=device)
model.eval()
```

The `.ckpt` file embeds hparams, so no separate config file is needed. This is the simplest approach and avoids Hydra instantiation complexity.

## Audio Preprocessing (Inlined in match.py)

The logic from `AudioFolderDataset.__getitem__` is inlined directly in `match.py`:

1. Load WAV via `pedalboard.io.AudioFile`, resample to 44.1kHz
2. Handle mono → stereo conversion (duplicate channel)
3. Pad/trim to 4 seconds with 50ms onset offset
4. Scale amplitude by 0.5
5. Compute mel-spectrogram via `librosa.feature.melspectrogram` (128 mels, AST-compatible)
6. Convert to dB scale
7. Normalize with stats.npz mean/std
8. Convert to float32 tensor, add batch dimension

## Constraints and Follow-ups

**Pedalboard preset export**: Pedalboard's `VST3Plugin` has `load_preset()` and parameter setters but no `save_preset()` method. V1 outputs JSON; `.vstpreset` export requires investigating the Surge XT preset binary format in a follow-up.

**Surge XT discovery**: Handled by an existing utility (out of scope for this spec). `apply_match.py` takes `--plugin-path` and trusts it.

**Surge XT installation**: Users install Surge XT themselves (free, open source). The tool does not bundle or auto-install it.

## Reusable Components

| Need | Function | File |
|------|----------|------|
| Mel-spectrogram | `make_spectrogram()` | `src/data/audio_datamodule.py` |
| Param decode | `ParamSpec.decode()` | `src/data/vst/param_spec.py` |
| Param specs registry | `param_specs` dict | `src/data/vst/__init__.py` |
| Load plugin | `load_plugin()` | `src/data/vst/core.py` |
| Load preset | `load_preset()` | `src/data/vst/core.py` |
| Set params | `set_params()` | `src/data/vst/core.py` |
| Render audio | `render_params()` | `src/data/vst/core.py` |
| Write WAV | `write_wav()` | `src/data/vst/core.py` |
| Model | `SurgeFlowMatchingModule` | `src/models/surge_flow_matching_module.py` |

## New Files

1. `scripts/match.py` — CLI + weight download + single-file inference glue
2. `scripts/apply_match.py` — CLI + JSON loading + Surge param application
3. Makefile targets: `match`, `apply`

## Verification

1. **Unit tests**: weight download logic (mock HTTP, verify cache paths), JSON output schema validation
2. **Integration test** (`@pytest.mark.slow`): full pipeline with real checkpoint + audio file, verify JSON contains expected param names from `SURGE_XT_PARAM_SPEC`
3. **Manual test**: `make match FILE=tests/fixtures/test.wav` → inspect JSON → `make apply FILE=test_match.json` → verify Surge loads params correctly
4. Apply command tests require Surge XT installed → also `@pytest.mark.slow`

**Comment by ktinubu** [comment]
## ⚠️ Implementation gotcha: mel-spectrogram param drift

While reviewing existing code, I found the repo has **two `make_spectrogram` functions with different parameters** that serve different purposes. `match.py` must use the correct one.

| Function | `n_fft` | `hop_length` | Purpose |
|----------|---------|--------------|---------|
| `src/data/audio_datamodule.py:make_spectrogram` | `int(0.025 * sr)` = 1102 | `int(sr / 100)` = 441 | **Training + inference** (AST-compatible) |
| `src/data/vst/generate_vst_dataset.py:make_spectrogram` | `int(0.025 * sr)` = 1102 | `int(sr / 100)` = 441 | **Dataset generation** (identical to above) |
| `scripts/predict_vst_audio.py:make_spectrogram` | 2048 | 512 | **Debug visualization only** (PNG plots) |

### Why they differ

- The **AST-compatible params** (1102 / 441) produce a mel-spec with shape `[128, 401]` for 4s of 44.1kHz audio (100 frames/sec × 4s + 1). These match what the Audio Spectrogram Transformer encoder was pretrained on.
- The **2048 / 512** variant in `predict_vst_audio.py` uses standard librosa defaults to produce nicer-looking PNG comparison plots for humans. Its output is never fed to the model.

### Implication for match.py

`match.py` **must** use the AST-compatible params (1102 / 441). Using the visualization params would produce a spectrogram with wrong time/frequency dimensions and the model would output garbage predictions.

Since we decided to inline audio preprocessing in `match.py` (no shared function extraction), the implementation will copy the logic from `audio_datamodule.py`, not `predict_vst_audio.py`.

### Follow-up (out of scope for this epic)

The AST-compatible `make_spectrogram` is currently duplicated between `audio_datamodule.py` and `generate_vst_dataset.py`. A future refactor should consolidate these into a single shared helper, but that's pre-existing drift, not something this epic introduces.
